### PR TITLE
Feature: multi-catalog

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

new feature: multiple catalogs

-  objects of catalog, e.g table, can be addressed as `catalog.database.table`
    only the current ut/it are covered, there are definitely other codes that should be modified to fully support this
-  all the objects that are kept in databend meta server belong to the `default` catalog  
-  new catalog can be registered by using the `CatalogManager`
- currently, a simple impl of `hive` catalog is provided, which can be registered by using de configs.
   - registered as catalog name `hive`
   - can resolve metadata by using hive meta store
   - a dummy implementation of hive database/table
      - only resolves simple and limit number of column types
      -  returns empty dataset for queries
- by default, the cargo feature `hive` of crate `databend-query` is NOT enabled
     - to enable it
       `cargo build --bin databend-query --features hive`
     - but the crate `common-hive-meta-store` will be built, even the `hive` feature is not specified
       - the only extra build time dependency is the `thrift` compiler which is built into dev docker images 
       - if there is a way to conditionally compile a number of workspace, pls let me know.
- A dedicate github action `test_stateful_hive_standalone` is added  
     -  only linux + amd64 are tested
     -  setups hive cluster (and simple test data)
 
    

## Changelog

- New Feature
- Not for changelog (changelog entry is not required)

## Related Issues

Fixes #issue

### CODE REVIEW

**Comment by dantengsky** [comment]
the files that `thrift` generated are HUGE. later,  they will be generated by "build.rs"

File: query/src/pipelines/transforms/transform_sink.rs [hunk 1]
```diff
@@ -84,7 +83,8 @@ impl Processor for SinkTransform {
         tracing::debug!("executing sinks transform");
         let tbl = self
             .ctx
-            .get_catalog()
+            //.get_catalog(&self.table_info.catalog_name)
+            .get_catalog("default")? // TODO pass this in
```
Comment (BohuTANG) [inline query/src/pipelines/transforms/transform_sink.rs:87]: get_catalog should not give the name, we should get them by some thing, such as the query plan, or it's complex to do it.

Comment (dantengsky) [reply query/src/pipelines/transforms/transform_sink.rs:87]: agree. PR is WIP,  and a TODO item is left there (line 87).

The name of catalog is planning to be passed in, maybe something like this:

https://github.com/datafuselabs/databend/blob/9fd413686832ad0654b9581a3cdf1d6ac59a773f/common/planners/src/plan_table_drop.rs#L25-L28

or  embedded in the SinkTransform::table_info

Comment (dantengsky) [reply query/src/pipelines/transforms/transform_sink.rs:87]: >  and set it to the session when parsing SQL

not sure about this, a session may need to work on serval catalogs simultaneously:

`select ... from default.db.tbl, hive.db.tbl ....`

Comment (dantengsky) [reply query/src/pipelines/transforms/transform_sink.rs:87]: Oh... got ur idea, for a given `sink`, it is supposed to be working on exact one `catalog`

Comment (BohuTANG) [reply query/src/pipelines/transforms/transform_sink.rs:87]: This is the point, maybe it helpful for us to simplify the logic.

Comment (BohuTANG) [reply query/src/pipelines/transforms/transform_sink.rs:87]: > agree
> 
> The name of catalog is planning to be passed in, maybe something like this:
> 
> https://github.com/datafuselabs/databend/blob/9fd413686832ad0654b9581a3cdf1d6ac59a773f/common/planners/src/plan_table_drop.rs#L25-L28
> 
> or embedded in the SinkTransform::table_info

👌

File: query/src/sql/statements/statement_common.rs [hunk 2]
```diff
@@ -198,14 +198,22 @@ pub fn resolve_table(
...
+            statement_name
+        ))),
         1 => Ok((ctx.get_current_catalog(), idents[0].value.clone())),
         2 => Ok((idents[0].value.clone(), idents[1].value.clone())),
-        _ => Err(ErrorCode::SyntaxException(
-            "database name must be [`catalog`].db`",
-        )),
+        _ => Err(ErrorCode::SyntaxException(format!(
+            "database name should be [`catalog`].`db` in statement {}",
+            statement_name
```
Comment (BohuTANG) [inline query/src/sql/statements/statement_common.rs:216]: We can get the query text from `self.ctx.get_query_str();`, it seems no need the `statement_name` here?

By the way, we also can trim the whole query to short using:
`SQLCommon::short_sql(query))`

Comment (BohuTANG) [reply query/src/sql/statements/statement_common.rs:216]: Oh, just found that the logic is already there not the new changed, please ignore :D

File: tests/suites/0_stateless/05_ddl/05_0001_ddl_create_database.sql [hunk 3]
```diff
@@ -12,7 +12,7 @@ DROP DATABASE IF EXISTS db;
 CREATE DATABASE system; -- {ErrorCode 2301}
 DROP DATABASE system; -- {ErrorCode 1002}
 
-CREATE DATABASE db.t; -- {ErrorCode 1005}
+CREATE DATABASE db.t; -- {ErrorCode 1006}
```
Comment (dantengsky) [inline tests/suites/0_stateless/05_ddl/05_0001_ddl_create_database.sql:15]: why?

File: tests/suites/0_stateless/05_ddl/05_0001_ddl_create_database.sql [hunk 4]
```diff
@@ -28,4 +28,4 @@ DROP SCHEMA IF EXISTS db;
 CREATE SCHEMA system; -- {ErrorCode 2301}
 DROP SCHEMA system; -- {ErrorCode 1002}
 
-CREATE SCHEMA db.t; -- {ErrorCode 1005}
+CREATE SCHEMA db.t; -- {ErrorCode 1006}
```
Comment (dantengsky) [inline tests/suites/0_stateless/05_ddl/05_0001_ddl_create_database.sql:31]: what's happening? why the err code changed?

File: query/tests/it/users/role_cache_mgr.rs [hunk 5]
```diff
@@ -31,7 +31,7 @@ async fn test_role_cache_mgr() -> Result<()> {
 
     let mut role1 = RoleInfo::new("role1");
     role1.grants.grant_privileges(
-        &GrantObject::Database("db1".to_string()),
+        &GrantObject::Database("default".to_string(), "db1".to_string()),
```
Comment (dantengsky) [inline query/tests/it/users/role_cache_mgr.rs:34]: use the constant value

**Review by BohuTANG — COMMENTED** [review]
Why do we delete common/ast/tests/it/ast.rs?

**Comment by dantengsky** [comment]
> Why do we delete common/ast/tests/it/ast.rs?

It was deleted(refactored) in PR https://github.com/datafuselabs/databend/pull/5089/files

**Comment by Xuanwo** [comment]
Hi, I have a question here: Are `hive` and `hdfs` separated? 

For example, users can:

- Use `hive` as metadata and store data inside `s3`?
- Use `fuse` as metadata and store data inside `hdfs`?

If so, our feature may be named `meta-hive`?

**Comment by dantengsky** [comment]
> Hi, I have a question here: Are `hive` and `hdfs` separated?

yes

> For example, users can:
> 
> * Use `hive` as metadata and store data inside `s3`?
> * Use `fuse` as metadata and store data inside `hdfs`?
> 

no，none of them

> If so, our feature may be named `meta-hive`?

File: query/src/storages/fuse/fuse_table.rs [hunk 6]
```diff
@@ -83,13 +95,72 @@ impl FuseTable {
         Ok(format!("{}/{}", db_id, table_id))
     }
 
-    pub fn description() -> StorageDescription {
-        StorageDescription {
-            engine_name: "FUSE".to_string(),
-            comment: "FUSE Storage Engine".to_string(),
-            support_order_key: true,
+    //    pub fn catalog_name(&self) -> Result<String> {
```
Comment (dantengsky) [inline query/src/storages/fuse/fuse_table.rs:98]: remove those

File: query/src/storages/fuse/fuse_table.rs [hunk 7]
```diff
@@ -163,16 +234,18 @@ impl Table for FuseTable {
 
     async fn commit_insertion(
         &self,
-        ctx: Arc<QueryContext>,
-        operations: Vec<DataBlock>,
-        overwrite: bool,
+        _ctx: Arc<QueryContext>,
```
Comment (dantengsky) [inline query/src/storages/fuse/fuse_table.rs:237]: rename these

File: query/src/storages/fuse/operations/commit.rs [hunk 8]
```diff
@@ -242,7 +242,8 @@ impl FuseTable {
         table_info: &TableInfo,
         new_snapshot_location: String,
     ) -> Result<UpsertTableOptionReply> {
-        let catalog = ctx.get_catalog();
+        // TODO catalog name
```
Comment (dantengsky) [inline query/src/storages/fuse/operations/commit.rs:245]: use arg `catalog_name`

File: query/src/storages/fuse/table_functions/fuse_snapshots/fuse_snapshot_table.rs [hunk 9]
```diff
@@ -200,7 +207,7 @@ impl AsyncSource for FuseHistorySource {
             let tenant_id = self.ctx.get_tenant();
             let tbl = self
                 .ctx
-                .get_catalog()
+                .get_catalog(&self.catalog_name)? // TODO pass in this guy
```
Comment (dantengsky) [inline query/src/storages/fuse/table_functions/fuse_snapshots/fuse_snapshot_table.rs:210]: clean up TODO

File: query/src/storages/memory/memory_table.rs [hunk 10]
```diff
@@ -263,22 +263,23 @@ impl Table for MemoryTable {
 
     async fn commit_insertion(
         &self,
-        ctx: Arc<QueryContext>,
-        operations: Vec<DataBlock>,
-        overwrite: bool,
+        _ctx: Arc<QueryContext>,
```
Comment (dantengsky) [inline query/src/storages/memory/memory_table.rs:266]: rename these

File: query/src/storages/system/columns_table.rs [hunk 11]
```diff
@@ -96,7 +95,8 @@ impl ColumnsTable {
         ctx: Arc<QueryContext>,
     ) -> Result<Vec<(String, String, DataField)>> {
         let tenant = ctx.get_tenant();
-        let catalog = ctx.get_catalog();
+        // TODO replace default with real cat
+        let catalog = ctx.get_catalog("default")?;
```
Comment (dantengsky) [inline query/src/storages/system/columns_table.rs:99]: use constant DEFAULT_CATALOG

File: query/src/storages/system/engines_table.rs [hunk 12]
```diff
@@ -21,25 +21,26 @@ use common_meta_types::TableIdent;
...
 
     fn get_table_info(&self) -> &TableInfo {
         &self.table_info
     }
 
-    fn get_full_data(&self, ctx: Arc<QueryContext>) -> Result<DataBlock> {
-        let table_engine_descriptors = ctx.get_catalog().get_table_engines();
+    async fn get_full_data(&self, ctx: Arc<QueryContext>) -> Result<DataBlock> {
+        // TODO passin catalog name
+        let table_engine_descriptors = ctx.get_catalog("default")?.get_table_engines();
```
Comment (dantengsky) [inline query/src/storages/system/engines_table.rs:43]: use the CONSTANT

File: query/src/storages/system/tables_table.rs [hunk 13]
```diff
@@ -41,7 +40,8 @@ impl AsyncSystemTable for TablesTable {
 
     async fn get_full_data(&self, ctx: Arc<QueryContext>) -> Result<DataBlock> {
         let tenant = ctx.get_tenant();
-        let catalog = ctx.get_catalog();
+        // TODO pass catalog in or embed catalog in table info?
+        let catalog = ctx.get_catalog("default")?;
```
Comment (dantengsky) [inline query/src/storages/system/tables_table.rs:44]: use the constant

File: query/tests/it/sql/statements/query/query_qualified_rewriter.rs [hunk 14]
```diff
@@ -42,61 +42,97 @@ async fn test_query_qualified_rewriter() -> Result<()> {
...
             name: "Alias query with aggregate",
             query: "SELECT COUNT(alias.name) AS name FROM system.databases AS alias WHERE name = 'xxx'",
             expect: "NormalQuery { filter: (name = xxx), aggregate: [COUNT(name)], projection: [COUNT(name) as name] }",
         },
-        TestCase {
-            name: "Database and table query with aggregate",
-            query: "SELECT COUNT(system.databases.name) AS name FROM system.databases WHERE system.databases.name = 'xxx'",
-            expect: "NormalQuery { filter: (name = xxx), aggregate: [COUNT(name)], projection: [COUNT(name) as name] }",
-        },
+//  TODO confirm this
```
Comment (dantengsky) [inline query/tests/it/sql/statements/query/query_qualified_rewriter.rs:130]: what's going on herer?

File: query/tests/it/storages/fuse/pruning.rs [hunk 15]
```diff
@@ -92,7 +93,7 @@ async fn test_block_pruner() -> Result<()> {
     interpreter.execute(None).await?;
 
     // get table
-    let catalog = ctx.get_catalog();
+    let catalog = ctx.get_catalog("default")?;
```
Comment (dantengsky) [inline query/tests/it/storages/fuse/pruning.rs:96]: use constant

File: query/tests/it/storages/fuse/pruning.rs [hunk 16]
```diff
@@ -229,7 +231,7 @@ async fn test_block_pruner_monotonic() -> Result<()> {
         order_keys: vec![],
     };
 
-    let catalog = ctx.get_catalog();
+    let catalog = ctx.get_catalog("default")?;
```
Comment (dantengsky) [inline query/tests/it/storages/fuse/pruning.rs:234]: use constant

File: query/tests/it/storages/fuse/table_test_fixture.rs [hunk 17]
```diff
@@ -100,6 +102,10 @@ impl TestFixture {
         gen_db_name(&self.prefix)
     }
 
+    pub fn default_catalog_name(&self) -> String {
+        "default".to_owned()
```
Comment (dantengsky) [inline query/tests/it/storages/fuse/table_test_fixture.rs:106]: use constant

File: tests/suites/0_stateless/05_ddl/05_0006_ddl_grant_privilege.result [hunk 18]
```diff
@@ -1,7 +1,7 @@
-GRANT ALL ON 'default'.* TO 'test-grant'@'localhost'
-GRANT SELECT ON 'db01'.* TO 'test-grant'@'localhost'
-GRANT SELECT ON 'db01'.'tb1' TO 'test-grant'@'localhost'
-GRANT ALL ON 'default'.* TO 'test-grant'@'localhost'
-GRANT SELECT ON 'db01'.'tb1' TO 'test-grant'@'localhost'
-GRANT SELECT ON 'db01'.'tb1' TO 'test-grant'@'localhost'
-GRANT SELECT ON 'default'.* TO 'test-grant-role'
+GRANT ALL ON 'default'.'default'.* TO 'test-grant'@'localhost'
```
Comment (dantengsky) [inline tests/suites/0_stateless/05_ddl/05_0006_ddl_grant_privilege.result:1]: append docs about privilege in the PR summary

File: common/hive-meta-store/build.rs [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    );
+    let mut output = File::create(output_file_path).expect("create output patch file failure");
+    for line in input.lines() {
+        let line = line.expect("readline failure");
+        if !line.starts_with("#![") {
+            std::writeln!(output, "{}", line).expect("write line to patched file failure");
+        }
+    }
+
+    println!("cargo:rerun-if-changed=build.rs");
```
Comment (Xuanwo) [inline common/hive-meta-store/build.rs:68]: Do we need to move this line to the front?

**Comment by BohuTANG** [comment]
@dantengsky 

LGTM.
I have seen your comments, I suggest refining them in another PR.
I will merge this monster PR first.

**Comment by BohuTANG** [comment]
The review also welcomed on this PR cc @andylokandy @Xuanwo

File: common/hive-meta-store/build.rs (hunk 19 above)
Comment (dantengsky) [reply common/hive-meta-store/build.rs:68]: Not sure about this : (
will moving this line to the front prevent us from some unexpected results? I am not quite familiar with the "mechanism" there.

Comment (Xuanwo) [reply common/hive-meta-store/build.rs:68]: Me too...

But most build.rs will place the println at the very first line. It's better to keep fit with them.

Comment (dantengsky) [reply common/hive-meta-store/build.rs:68]: got it & thanks. I will fix it in PR #5284

**Pull request merged**